core/ir/instr_shared.c

```diff
@@ -3697,7 +3697,8 @@ instr_is_reg_spill_or_restore_ex(void *drcontext, instr_t *instr, bool DR_only,
     if (reg == NULL)
         reg = &myreg;
     if (instr_check_tls_spill_restore(instr, spill, reg, &check_disp)) {
-        if (!DR_only ||
+        /* We do not want to count an mcontext base load as a reg spill/restore. */
+        if ((!DR_only && check_disp != os_tls_offset((ushort)TLS_DCONTEXT_SLOT)) ||
```

Comment:
Besides drreg this path is used by drx during scatter/gather state restoration, and I don't think that block contains mcontext base loads.

It's also used in core/translate.c

https://github.com/DynamoRIO/dynamorio/blob/1cf4a13ff5f2acf111f45edd8f5d781421f070ff/core/translate.c#L598

to assert `!spill`. For mcontext base, I think we'd always do a load, not a store, so it'd never fail that assert.

Reply:
> this path is used by drx during scatter/gather state restoration, and I don't think that block contains mcontext base loads.

Actually some client may instrument a clean call which does add mcontext base loads.

Reply:
I think drx's state restoration for scatter/gather may need some work. Besides confusion due to not being able to recognise mcontext base loads (after this PR), there's also the existing issue that loads using the mcontext (restores from mcontext slots) might mess up the state transitions. There's already a sort of TODO in drx to fix issues caused by client instrumentation. I've created #5005 to investigate this and make it more robust. Also added it to drx.

Reply:
Just want to confirm my understand: when `!DR_only`, on x86 the only additional offset we return true for is `TLS_DCONTEXT_SLOT` right?https://github.com/DynamoRIO/dynamorio/blob/51c67c8e0e88af20ea19732cdb27cecff8df73d7/core/arch/arch_exports.h#L147

Why is the arg called `DR_only`? Does it mean that core DR will never load from `TLS_DCONTEXT_SLOT`?

Reply:
> Just want to confirm my understand: when !DR_only, on x86 the only additional offset we return true for is TLS_DCONTEXT_SLOT right?

No, for !DR_only it would return true for any offset that uses DR's segment register: so any raw client slot (hmm, on Windows it would include app TEB accesses of regular TEB fields, and similarly on Mac?  Or is there an offset range check for those shared-TLS setups?).

> Why is the arg called DR_only?

"DR_only" means it only considers the scratch slots that DR owns and not slots handed out to clients.

> Does it mean that core DR will never load from TLS_DCONTEXT_SLOT?

I think this was just never needed to be recognized when restoring DR's own spills.

Reply:
> I think this was just never needed to be recognized when restoring DR's own spills.

Yeah, atleast `dr_save_reg` and `dr_restore_reg` restore the tmp reg where mcontext base was loaded.

I'll verify and fix the drx state restoration in #5005.